src/lib.rs

```diff
@@ -41,1346 +53,988 @@ use std::{
+        let mut eval_points = std::collections::HashSet::new();
+
+        // Collect middleware results into vectors
+        let mut x_polynomials = vec![];
+        let mut polynomials = vec![];
+        let mut randomnesses = vec![];
+
+        for (label, (_point_label, point)) in query_set.iter() {
+            eval_points.insert(*point);
+
+            if !commitment_map.contains_key(label) {
+                end_timer!(combine_time);
+                Err(Error::MissingCommitment {
+                    label: label.to_string(),
+                })?
             }
 
-            let supported_hiding_bound = polynomials
-                .iter()
-                .map(|p| p.hiding_bound().unwrap_or(0))
-                .max()
-                .unwrap_or(0);
-            println!("supported degree: {:?}", supported_degree);
-            println!("supported hiding bound: {:?}", supported_hiding_bound);
-            let (ck, vk) = PC::trim(&pp, supported_degree)?;
-            assert!(ck.is_valid());
-            assert!(vk.is_valid());
-            println!("Trimmed");
-
-            test_canonical_serialize_deserialize(true, &ck);
-            test_canonical_serialize_deserialize(true, &vk);
-
-            let (comms, rands) = PC::commit(&ck, &polynomials, Some(rng))?;
-            assert!(comms.is_valid());
-
-            let mut query_set = QuerySet::new();
-            let mut values = Evaluations::new();
-            let point = F::rand(rng);
-            for (i, label) in labels.iter().enumerate() {
-                query_set.insert((label.clone(), (format!("{}", i), point)));
-                let value = polynomials[i].evaluate(point);
-                values.insert((label.clone(), point), value);
+            let labeled_polynomial = *poly_map.get(label).ok_or(Error::MissingPolynomial {
+                label: label.to_string(),
+            })?;
+
+            let labeled_randomness = *rands_map.get(label).ok_or(Error::MissingRandomness {
+                label: label.to_string(),
+            })?;
+
+            if labeled_polynomial.is_hiding() {
+                has_hiding = true;
             }
-            println!("Generated query set");
-
-            let mut fs_rng = setup_test_fs_rng::<F, PC>();
-            let proof = PC::batch_open(
-                &ck,
-                &polynomials,
-                &comms,
-                &query_set,
-                &mut fs_rng,
-                &rands,
-                Some(rng),
-            )?;
-            assert!(proof.is_valid());
 
-            test_canonical_serialize_deserialize(true, &proof);
+            // y_i
+            let y_i = labeled_polynomial.polynomial().evaluate(*point);
+
+            // (X - x_i)
+            let x_polynomial =
+                Polynomial::from_coefficients_slice(&[-(*point), G::ScalarField::one()]);
+
+            // z(X) = PROD (X - x_i)
+            z_polynomial = z_polynomial.naive_mul(&x_polynomial);
```

Comment:
Regardless of the fact that we do not use `z(X)` anymore: The correct computation of it needs to assure that query points are not taken into account twice.